Hub.IsEnabled set to false when Hub disposed

semuserable opened the pull request:
`Hub.IsEnabled` is set to `false` when `Hub.Dispose()` is called. Related to https://github.com/getsentry/sentry-unity/issues/193 bug.

Tyrrrz commented on src/Sentry/Internal/Hub.cs:
```diff
@@ -258,6 +260,7 @@ public void Dispose()
                 }
             }
 
+            Interlocked.Exchange(ref _isEnabled, 0);
```
Does this helps us in any way? I'm thinking because by this point we're already half way into disposing anyway, so if there was a race condition, it could have occurred already. Maybe just an `bool _isDisposed` field is enough? Which scenarios are we trying to protect against here?

semuserable replied:
Tbh, this is what we discussed with Bruno. My initial thought was the same, just `bool`, but Bruno insisted on `Interlocked`. From my understanding, the main scenario for this is that `Dispose` can be called by more than 1 thread.

bruno-garcia replied:
> but Bruno insisted on Interlocked

I don't recall a conversation where I insisted. 
What I recall is that I suggested a solution with `Interlocked` and there was no discussion of trade off.

bruno-garcia commented on src/Sentry/Internal/Hub.cs:
```diff
@@ -258,6 +260,7 @@ public void Dispose()
                 }
             }
 
+            Interlocked.Exchange(ref _isEnabled, 0);
```
It's just missing the check to avoid double dispose:

```suggestion
            if (Interlocked.Exchange(ref _isEnabled, 0) != 0)
            {
                return; // Already disposed
            }
```

bruno-garcia replied:
But this must be the first thing (maybe after logging)

semuserable replied:
Oh, I see.

bruno-garcia commented:
Also, could you please add a test to validate that it in fact changes from true to false once u dispose?

semuserable commented:
Updated. `Interlocked` moved up; Unit test added.

@bruno-garcia I added a test with `InlineData` to potentially check "over-disposing". I'm not sure it's needed for now, but still.

bruno-garcia commented on test/Sentry.Tests/HubTests.cs:
```diff
@@ -639,5 +639,26 @@ public void CaptureTransaction_AfterTransactionFinishes_ResetsTransactionOnScope
+        public void Dispose_IsEnabled_SetToFalse(int disposeCount)
+        {
+            // Arrange
+            var hub = new Hub(new SentryOptions
+            {
+                Dsn = DsnSamples.ValidDsnWithSecret
+            });
+
+            // Act
+            for (var i = 0; i < disposeCount; i++)
+            {
+                hub.Dispose();
+            }
+
+            // Assert
+            hub.IsEnabled.Should().BeFalse();
```
As discussed on our call:

```suggestion
            var hub = new Hub(new SentryOptions
            {
                Dsn = DsnSamples.ValidDsnWithSecret,
                
            });
            var client = Substibute.For<ISentryClient>();
            hub.BindClient(client);
            // Act
            for (var i = 0; i < disposeCount; i++)
            {
                hub.Dispose();
            }
            client.Received(1).Dispose();
            // Assert
            hub.IsEnabled.Should().BeFalse();
```

Another test would verify that `IsDispose` returns true at first, and after Dispose is called, it returns false

semuserable commented:
updated

cc @bruno-garcia @Tyrrrz
